### ISSUE TITLE
Optimize seeks to end

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#411]
Stream implementations in quite a few languages do not have a simple way to get a size of the stream, but allow to seek to a position relative to end of stream. This can be used to optimize clauses like `pos: _io.size - 123`. If a language supports "seeking relative to end of stream", instead of doing two (or even more) calls, we can just translate that into one `seek_from_end(-123)`.

Given that it stems from C, where `int fseek(FILE *stream, long offset, int whence)` has `whence` that allows to seek relative to end of file, it is very widely available.

This would be useful at least for:

* [C++](https://github.com/kaitai-io/kaitai_struct_cpp_stl_runtime/blob/master/kaitai/kaitaistream.cpp#L87)
* [C#](https://msdn.microsoft.com/en-us/library/system.io.filestream.seek(v=vs.110).aspx)
* [Go](https://github.com/kaitai-io/kaitai_struct_go_runtime/blob/master/kaitai/stream.go#L57)
* [Perl](http://perldoc.perl.org/functions/seek.html)
* [PHP](http://php.net/manual/en/function.fseek.php)
* [Python](https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/master/kaitaistruct.py#L91)
* [Ruby](https://ruby-doc.org/core-2.2.2/IO.html#method-i-seek)

Probably won't be useful for:

* Java — lacks relevant seek-relative-to-end API
* JS — we do all offset arithmetic manually anyway
* Lua — seems to [lack API to seek *relatively* to end of file](https://www.lua.org/pil/21.3.html)

Implementing these require 2 steps:

* Add relevant runtime API like `seek_from_end(offset)` — anyone can come up with a better name?
* Update ksc to detect and apply this optimization where possible:
  * when `pos` employs `_io.size` and `io` is default `_io`
  * when `pos` employs `some_io.size` and `io` is set to that `some_io`

**Comment by arekbulski** [comment]
I hate to shoot down ideas for optimisation but feel oblidged to point out: What you are saving there is literally _1 integer addition_, right? You know how cheap that is? Even on Python?

There is a related optimisation, similar in performance gain, but easier to implement: The seek function, in both C and also confirmed in C# Python Lua, as you can see in the signature, returns current offset. In the Python runtime ([see code](https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/5dfb881636c86f4e6a7778b7c513174d8ccfffe2/kaitaistruct.py#L90-L93)) you have `seek(); x=tell()` which can be replaced by `x = seek()`. I will update Python runtime to that end (forgive the pun). C# runtime does not use seek for sizing, not applicable.

Lua page would suggest `file:seek("end", -offset)` as valid code.

**Comment by arekbulski** [comment]
There is also another optimisation: You can make a resonable assumption that the parsed file remains frozen throught parsing, therefore you can measure size once (in ctor) and just use a cached/precomputed value instead of seeking each time. 

Python PR will follow shortly.

**Comment by GreyCat** [comment]
> What you are saving there is literally 1 integer addition, right?

No, I'm saving a call to `_io.size` here, which is frequently implemented as:

* "Remember current position" => on Linux typically goes through `ftell` in glibc => `fseek` => `lseek`, and a mandatory lock on file descriptor mutex) => ~2-3 syscalls
* "Set position to end of file" => on Linux typically goes through `fseek` => `lseek`, and another set of locks => ~2-3 syscalls
* If we're unlucky and that "set position" discarded value from `lseek`, we'll have to call yet another `ftell` => ~2-3 more syscalls
* "Set position back to remembered" => again a `fseek` and ~2-3 more syscalls

This is actually a lot of syscalls that are just wasted.

Also, I've just tried Python and it can't do positioning relative to end, only directly *to* the end of stream:

    UnsupportedOperation: can't do nonzero end-relative seeks

Sob-sob.

**Comment by GreyCat** [comment]
Did a quick Ruby benchmark. <details>

```ruby
#!/usr/bin/env ruby

require 'benchmark'

f = File.open('/etc/passwd', 'r')
n = 10000000

Benchmark.bm { |bm|
  bm.report('direct seek') {
    n.times {
      f.seek(-5, IO::SEEK_END)
    }
  }
  bm.report('size + seek') {
    n.times {
      f.seek(f.size - 5)
    }
  }
  bm.report('emulate lack of size') {
    n.times {
      # _io.size
      old_pos = f.pos
      f.seek(0, IO::SEEK_END)
      file_size = f.pos
      f.seek(old_pos)

      # finally, a seek
      f.seek(file_size - 5)
    }
  }
}
```

</details>


Results on my box:

| | user | system | total | real |
|-|-|-|-|-|
| direct seek | 2.169195 | 1.764233 | 3.933428 | (  3.934675) |
| size + seek | 3.756040 | 4.056105 | 7.812145 | (  7.815489) |
| emulate lack of size | 9.192413 | 8.696388 | 17.888801 | ( 17.892883) |

I.e. the proposed optimization is ~4.5x faster than current "dumb" approach.

**Comment by GreyCat** [comment]
What am I doing wrong?

```
Python 3.6.5rc1 (default, Mar 14 2018, 06:54:23) 
...
In [1]: import io

In [2]: f = open('/etc/passwd')

In [3]: f.seek(-5, io.SEEK_END)
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-3-b80043a069f9> in <module>()
----> 1 f.seek(-5, io.SEEK_END)

UnsupportedOperation: can't do nonzero end-relative seeks
```

**Comment by arekbulski** [comment]
```py3
Python 3.6.3 (default, Oct  3 2017, 21:45:48) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> f = open('temp1','w+b')
>>> f.write(bytes(100))
100
>>> f.seek(0)
0
>>> f.seek(-5,io.SEEK_END)
95
```
Maybe that file has something that makes it "less seekable"?... :confused:

**Comment by arekbulski** [comment]
updated last post

**Comment by GreyCat** [comment]
Oh, my bad, thanks! Then we can actually do the benchmark on Python and see the benefits there as well.

**Comment by arekbulski** [comment]
Just in case you (or whoever else is also reading this) missed it, I will repeat previous findings: I need to withdraw the 2 suggestions I made.
- Python file.seek returns current offset only on Python 3. It just returns None on Python 2. Unless you want to have separate runtimes, or use some hacks...
- Caching/Precomputing file size breaks functionality if a file is not frozen, which as you kindly pointed out, would be the case for growing system log files. I already removed those commits from PRs.

**Comment by arekbulski** [comment]
Ah, it took me an hour of staring at your Ruby benchmark code to understand what this optimisation is about. What you point out is that tag like instance `pos: _io.size - 123` does not translate to `seek(-123, END)` but to `seek( size() -123 )` where size() is either filestream property (on C#) or made of several seeks and tells (on Python). On C# and Python the first code would work, on Lua we need a workaround.

**Comment by arekbulski** [comment]
Perhaps the suggestions I made arent stupid after all. We could use precomputing stream size, not on the original stream, but on substreams, which are (1) frozen after all and (2) BytesIO which makes computing size a bit of code. This is a Python-only suggestion.

**Comment by arekbulski** [comment]
updated last post

**Comment by arekbulski** [comment]
Perhaps instead of adding `seek_from_end` it might be better to use the already exposed `seek` method with a whence=END parameter?
